### DISCUSSION TITLE
Creating timelines for blocks with different trial numbers

### DISCUSSION THREAD

**Discussion by ShengjieXu667** [jspsych/jsPsych#1810] [Q&A]
Hi all,

I am working on an jsPsych experiment code where I plan to use blocks with different trial numbers. There will be 4 sessions, each session has 8 blocks and 28-32 trials per block (the mean is 30). So, total trial number in each session is 240, but each block has different length. Besides, the trial number for each block is randomized, which means the difference among blocks is also different among subjects. As I noticed, in timeline variables, the repetitions parameter can't be a function. What I did is, I created a variable containing the trial numbers for each block of the current subjects, then extracted the corresponding value to pass to the repetitions parameter. But the result was, the first block was as expected, the second block would have the same length as the first block. It seems I can't do dynamical assignment to repetitions parameter in timeline variable. 

So in my situation, what should I do to realize this design? Should I have to write timeline variables for each block, and then put them together to build a nested timeline variable? If so, I have to build up 32 timeline variables and 4 nested timeline variables and 1 nested nested timeline variable. However, I guess there might be some more elegant way to solve this problem. Thanks in advance!

Best,
Shengjie

**Comment by cslongman** [comment]
Hi Shengjie,

Personally, I prefer to write a separate piece of code that does all of my randomisation etc. and stores everything I need to run the experiment in a huge array (that I usually call TrialDat). Within this array, I can have several variables (e.g., TrailDat.TrialNum, TrialDat.Stimulus, TrialDat.CorResp, etc...) and each variable can itself be an array of data that refers to each trial within each block (e.g., TrialDat.TrialNum[bl][tr]).

This way, I can separate out the two main parts of the program - one part organises the trial data into a nice coherent array, and another that then uses the data from the array to piece together the experiment using the timeline variables.

It sounds like this might be a good way forward for you.

In the code I am working on at the moment, I have 24 blocks, each block is made up of 12 miniblocks, and each miniblock has 2-7 trials in it. By storing all of the relevant data in one place, I can then use for loops to tell jsPsych which timeline variables to use, how many times to use them, and what order to put them in.

I am more than happy to share my code with you if you think it woudl help.

Cai

**Reply by siqicao** [reply]
Hi Cai,

After reading your reply, I supposed that my experiment could benefit from your design logic. However, I met some problems when programming. I would be grateful if you are willing to share your codes. Thank you!

**Reply by cslongman** [reply]
Hi @sylviacao,

Yes, I'm happy to share my code. I created a new GitLab repository for it here...

https://gitlab.pavlovia.org/cslongman/aeis_context_intermixed

If you look at experiment.html you can see the entire code (note that this particular experiment is not designed to run via Pavlovia).

Lines 24-501 are where I create the array I mentioned. I do this in a function that is called in line 504 where it populates the relevant variables/arrays.

The rest of the code is essentially putting the timeline together by running several for loops and adding timeline variables one at a time.

I am very happy to talk you through it in a bit more detail if you think it would help.

Cai

**Reply by siqicao** [reply]
Hi Cai, @cslongman 
I appreciate your instant response! : ) Unfortunately, clicking on the link turns out to be a '404' error. I am not sure if this is a private link or not. Anyway, is there any other chance that I can access it? Thank you a lot.

**Reply by cslongman** [reply]
Hi @sylviacao,

Sorry, I am kinda new to all of this too!

I tried a few things to change permissions, but I think I just ended up making matters worse! I ended up deleting the project from GitLab so that link will no longer work at all.

However, I have added my code to your other thread here...

[(https://github.com/jspsych/jsPsych/discussions/1884)]

Cai

**Comment by becky-gilbert** [comment]
Hi Shengjie, I think you should be able to use variables for your repetitions parameters, as long as each block is defined in its own timeline node. Here's a simplified example:
```js
// randomize the assignment of trial numbers to blocks
var block_repetitions = jsPsych.randomization.shuffle([28, 32]);

// define trial objects
// ...

//set up each block with the randomized repetition values

var block1 = {
  timeline: [trial],
  repetitions: block_repetitions[0]
};

var block2 = {
  timeline: [trial],
  repetitions: block_repetitions[1]
};
```
Does that work in your case?
